sdks/python/apache_beam/transforms/userstate.py

```diff
@@ -0,0 +1,200 @@
+# limitations under the License.
+#
+
+"""User-facing interfaces for the Beam State and Timer APIs.
+
+Experimental; no backwards-compatibility guarantees.
+"""
+
+from __future__ import absolute_import
+
+import logging
+
+from apache_beam.coders import Coder
+from apache_beam.transforms.timeutil import TimeDomain
+
+
+class StateSpec(object):
+  """Specification for a user DoFn state cell."""
+
+  def __init__(self):
+    raise NotImplementedError
+
+  def __repr__(self):
+    return '%s(%s)' % (self.__class__.__name__, self.name)
+
+
+class BagStateSpec(StateSpec):
+  """Specification for a user DoFn bag state cell."""
+
+  def __init__(self, name, coder):
+    assert isinstance(name, str)
+    assert isinstance(coder, Coder)
+    self.name = name
+    self.coder = coder
+
+
+class CombiningValueStateSpec(StateSpec):
+  """Specification for a user DoFn combining value state cell."""
+
+  def __init__(self, name, coder, combiner):
+    # Avoid circular import.
+    from apache_beam.transforms.core import CombineFn
+
+    assert isinstance(name, str)
+    assert isinstance(coder, Coder)
+    assert isinstance(combiner, CombineFn)
+    self.name = name
+    self.coder = coder
+    self.combiner = combiner
+
+
+class TimerSpec(object):
+  """Specification for a user stateful DoFn timer."""
+
+  def __init__(self, name, time_domain):
+    self.name = name
+    if time_domain not in (TimeDomain.WATERMARK, TimeDomain.REAL_TIME):
+      raise ValueError('Unsupported TimeDomain: %r.' % (time_domain,))
+    self.time_domain = time_domain
+
+  def __repr__(self):
+    return '%s(%s)' % (self.__class__.__name__, self.name)
+
+
+# Attribute for attaching a TimerSpec to an on_timer method.
+_TIMER_SPEC_ATTRIBUTE_NAME = '_on_timer_spec'
+
+
+def on_timer(timer_spec):
+  """Decorator for timer firing DoFn method.
+
+  This decorator allows a user to specify an on_timer processing method
+  in a stateful DoFn.  Sample usage:
+
+  > class MyDoFn(DoFn):
+  >   TIMER_SPEC = TimerSpec('timer', TimeDomain.WATERMARK)
+  >
+  >   @on_timer(TIMER_SPEC)
+  >   def my_timer_expiry_callback(self):
+  >     logging.info('Timer expired!')
+  """
+
+  if not isinstance(timer_spec, TimerSpec):
+    raise ValueError('@on_timer decorator expected TimerSpec.')
+
+  def _inner(method):
+    if not callable(method):
+      raise ValueError('@on_timer decorator expected callable.')
+    setattr(method, _TIMER_SPEC_ATTRIBUTE_NAME, timer_spec)
+    return method
+
+  return _inner
+
+
+class UserStateUtils(object):
+
+  @staticmethod
+  def get_on_timer_methods(dofn):
+    # Avoid circular import.
+    from apache_beam.transforms.core import DoFn
+    if not isinstance(dofn, DoFn):
+      raise ValueError('Expected DoFn.')
+
+    timerid_to_methods = {}
+    result = {}
+
+    for name in dir(dofn):
+      value = getattr(dofn, name, None)
+      if (callable(value) and
+          getattr(value, _TIMER_SPEC_ATTRIBUTE_NAME, None) is not None):
```

Comment:
I think it'd be cleaner to attach the method to the timer spec, rather than the other way around. This would also allow for better error reporting for incorrect use (e.g. raising an exception at the time of decoration).